### DISCUSSION TITLE
Storage for mods (and userspace directory structure)

### DISCUSSION THREAD

**Discussion by Dliwk** [efroemling/ballistica#235] [Ideas]
Currently there are only one 'cross-platform' location for storage some of mod files it needs, which is the mods folder.

It's probably a good idea to discuss the options for folder structures with mods data here.

**Comment by Dliwk** [comment]
I suggest something similar to this
```py <!-- yeah, very good method to color tree -->
BombSquad
├── data
│   └── awesomemod
│       └── databases
│           └── awesome.db
├── mods
│   └── awesomemod
│       ├── audio
│       │   └── awesomesound.ogg
│       ├── data
│       │   ├── languages
│       │   │   └── english.json
│       │   └── maps
│       │       └── awesome_map.json
│       ├── models
│       │   └── awesomemodel.bob
│       ├── python
│       │   └── awesomemod
│       │       └── __init__.py
│       └── textures
│           └── awesometexture.dds
└── python
    └── awesome_mod_single_file.py
```

- There is still directory for simple (or not) python modules - `python`.
- Static data stored in `mods` directory, so *theoretically* mods can be distributed in .zip/.tar.*/etc. formats.
- Some mod configs/databases/etc. stored in `data` directory, each mod chooses what to store and how to store itself.

**Reply by Dliwk** [reply]
Ok; if all assets are stored and managed by the engine, then

```py
BombSquad
├── data
│   └── awesomemod
│       ├── config.json
│       └── databases
│           └── awesome.db
└── mods
    └── awesomemod
        └── __init__.py
```

About `awesomemod/config.json`: maybe it makes sense to separate 'AppConfig' and 'ModConfig', the second can be fun to use
```python
# awesomemod/__init__.py

...

config = ba.ModConfig(__name__)

# Access it via
awesomemod.config['key'] = 'value'
```

**Comment by efroemling** [comment]
A few notes from my end:
Assets is going to be the major focus of 1.7. (I'm aaaaalmost done with 1.6 and then I'll switch my focus to that).

My longer term vision for the game is to have as much as possible be cloud based, and the first pieces of that will be asset-packages in 1.7.

Basically, modders will be able to upload asset source files (.wav, .png, etc) and group them into packages associated with their accounts and then any game scripts referring to those packages will result in the engine automatically downloading the correct compiled texture/sound/etc. files for the current platform. This means it'll be possible to run servers using mods and everyone connecting will be able to see everything without having to manually install anything.

I've already got the cloud-based asset build system working and just need to finish adding support to the game to download and use them. The changes to game scripts will be something like [this](https://github.com/efroemling/ballistica/wiki/Dependency-System) if you want to get a preview.

I'm sure I'll make some tutorials or more documentation as 1.7 nears and things start working. I'd love to get feedback from modders when that happens.

Note that this will only cover asset files such as textures, sounds, etc. that need to exist on all clients. Things like scripts will never be automatically downloaded for obvious security reasons, and some data files such as databases would only need to exist on the host. So perhaps this conversation can continue regarding those sorts of things...

**Reply by rahulraman0108** [reply]
I would love to use the 1.7 version with the Dependency System when it is released. And as for `engine automatically downloading the correct compiled texture/sound/etc. files` I would like it if there were more information like the following:

Will the engine download the assets synchronously or asynchronously?
If the assets download is started on the main game thread then the player will have to wait for all download to complete before they can do anything, or if it is started on another thread or async then the main game thread would continue working and the player might be able to join the server and get the dependency texture asset for example in the first second but it might not have the assets installed which would result probably in a similar way to being removed from the server with error.

So if currently possible, I would like to know how is the dependency downloading and server joining going to be handled?
It may work like this: The player tries to join a server gets a prompt to download dependency assets and a window stating to wait till the download is complete in the background thread before joining that server?
I am just guessing how it would work because I would like to use the Dependency System on servers whenever it completely rolls out and also to do debugging and testing with it.

**Comment by MatiasMGz** [comment]
So... any updates on this? I hope this idea gets implemented in 1.8.
